### PR TITLE
Rewrite Path to use native path character type

### PR DESCRIPTION
**Pull request by wilhuff**
Paths now hold their platform-specific string values. On Windows, portable pathnames are UTF-16 stored in `wchar_t` arrays, but there's no `absl::wstring_view`, so filesystem APIs can't be designed to operate on `absl::string_view`.

Also, using `string_view` doesn't really save any copies in practice. For example, I need `Dirname` to create directories recursively and that needs a null-terminated string for each `mkdir` call, forcing the caller to copy.

Another option would be to have both `Path` and `PathView`. This would avoid all unnecessary copies at the expense of a more complicated implementation. Since we don't use the filesystem that intensively I'm okay with just keeping this simpler.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/util/path.h [hunk 1]
```diff
@@ -20,37 +20,129 @@
...
+
+  const char_type* c_str() const {
+    return pathname_.c_str();
+  }
+
+  size_t size() const {
+    return pathname_.size();
+  }
+
+#if defined(_WIN32)
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/util/path.h]: Very optional: I understand that these functions are grouped logically. I would _personally_ find it easier to grasp if instead they were grouped by platform, which would minimize the number of `ifdef`s and allow to quickly see platform-specific interface. But it's a matter of preference, so if you find existing version more readable, or don't have a preference either way, then please just leave as is.

File: Firestore/core/src/firebase/firestore/util/path.cc [hunk 2]
```diff
@@ -16,96 +16,225 @@
...
+#if defined(_WIN32)
+Path::string_type Path::CanonicalPath() const {
+  if (pathname_.empty()) {
+    return pathname_;
+  }
+
+  // Remove trailing separators, but take care not to remove the trailing slash
+  // in a root path name (which can have a drive letter).
+  std::wstring copy{pathname_};
+  size_t min = StripDriveLetter(copy.c_str(), copy.size());
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/util/path.cc]: I couldn't understand what `min` stands for without reading the code. Perhaps something like `relative_path_begin`, `rel_path_first`, etc.?

File: Firestore/core/src/firebase/firestore/util/path.cc [hunk 3]
```diff
@@ -16,96 +16,225 @@
...
+    }
+  }
+  return Path::npos;
+}
+
+size_t LastSeparator(const Path::char_type* path, size_t size) {
+  if (size == 0) return Path::npos;
+
+  size_t i = size;
+  for (; i-- > 0;) {
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/path.cc]: I moved the initialization in. `i-- > 0` is the idiomatic way of looping through an unsigned int backward. I don't think conversion to signed is worthwhile.

File: Firestore/core/src/firebase/firestore/util/path.cc [hunk 4]
```diff
@@ -16,96 +16,225 @@
...
-static constexpr absl::string_view::size_type npos = absl::string_view::npos;
+namespace {
 
-/** Returns the given path with its leading drive letter removed. */
-inline absl::string_view StripDriveLetter(absl::string_view path) {
+/**
+ * Returns the offset within the given path that skips the leading drive letter.
+ * If there is no drive letter, returns zero.
+ */
+size_t StripDriveLetter(const Path::char_type* path, size_t size) {
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/util/path.cc]: I certainly prefer `OffsetAfterDriveLetter`.

File: Firestore/core/src/firebase/firestore/util/path.cc [hunk 5]
```diff
@@ -16,96 +16,225 @@
...
 
   } else {
-    size_t nonslash = base->find_last_not_of('/');
-    if (nonslash != npos) {
-      base->resize(nonslash + 1);
-      base->push_back('/');
+    size_t non_slash = LastNonSeparator(pathname_.c_str(), pathname_.size());
+    if (non_slash != npos) {
+      pathname_.resize(non_slash + 1);
+      pathname_.push_back(kPreferredSeparator);
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/util/path.cc]: Hmm, if that's a valid path on Windows, then perhaps we could just always use the forward slash and do away with `kPreferredSeparator` (the only reference in `CanonicalPath` could be inlined)? OTOH, it's probably not worth the trouble, and perhaps `kPreferredSeparator` will get more usage later on.

File: Firestore/core/src/firebase/firestore/util/path.cc [hunk 6]
```diff
@@ -16,96 +16,225 @@
...
+    return pathname_;
+  }
+
+  size_t last_slash = non_slash + 1;
+  return absl::string_view{c_str(), last_slash};
+}
+#endif  // defined(_WIN32)
+
+bool operator==(const Path& lhs, const Path& rhs) {
+  return lhs.CanonicalPath() == rhs.CanonicalPath();
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/util/path.cc]: Thanks for explaining.

**Pull request merged**